cpp/oneapi/dal/backend/primitives/sort/sort_dpc.cpp

```diff
@@ -0,0 +1,529 @@
+                                                  local_hist_count_,
+                                                  scan_deps);
+            res_deps = radix_reorder(queue_,
+                                     val_buff_,
+                                     ind_buff_,
+                                     part_prefix_hist_,
+                                     val_in,
+                                     ind_in,
+                                     elem_count_,
+                                     bit_offset,
+                                     local_size_,
+                                     local_hist_count_,
+                                     hist_scan_deps);
+        }
+    }
+
+    sort_event_ = res_deps;
+    return res_deps;
+}
+
+template <typename Integer>
+radix_sort<Integer>::radix_sort(const sycl::queue& queue, std::int64_t vector_count)
+        : queue_(queue),
+          vector_count_(0) {
+    init(queue_, vector_count);
+}
+
+template <typename Integer>
+radix_sort<Integer>::~radix_sort() {
+    sort_event_.wait_and_throw();
+}
+
+template <typename Integer>
+void radix_sort<Integer>::init(sycl::queue& queue, std::int64_t vector_count) {
+    std::uint32_t uint_vector_count = de::integral_cast<std::uint32_t>(vector_count);
+    if (uint_vector_count && vector_count_ != uint_vector_count) {
+        vector_count_ = uint_vector_count;
+
+        buffer_ = ndarray<Integer, 2>::empty(queue_,
+                                             { vector_count_, radix_range_ },
+                                             sycl::usm::alloc::device);
+    }
+}
+
+template <typename Integer>
+sycl::event radix_sort<Integer>::operator()(ndview<Integer, 2>& val_in,
+                                            ndview<Integer, 2>& val_out,
+                                            std::int64_t sorted_elem_count,
+                                            const event_vector& deps) {
+    // radixBuf should be big enough to accumulate radix_range elements
+    ONEDAL_ASSERT(val_in.get_dimension(0) > 0);
+    ONEDAL_ASSERT(val_in.get_dimension(0) < de::limits<std::uint32_t>::max());
+    ONEDAL_ASSERT(val_in.get_dimension(1) > 0);
+    ONEDAL_ASSERT(val_in.get_dimension(1) < de::limits<std::uint32_t>::max());
+    ONEDAL_ASSERT(val_in.get_dimension(0) == val_out.get_dimension(0));
+    ONEDAL_ASSERT(val_in.get_dimension(1) == val_out.get_dimension(1));
+    ONEDAL_ASSERT(val_out.has_mutable_data());
+    ONEDAL_ASSERT(sorted_elem_count > 0);
+    ONEDAL_ASSERT(sorted_elem_count <= val_in.get_dimension(1));
+
+    sort_event_.wait_and_throw();
+
+    const std::uint32_t vector_count = de::integral_cast<std::uint32_t>(val_in.get_dimension(0));
+    const std::uint32_t vector_offset = de::integral_cast<std::uint32_t>(val_in.get_dimension(1));
+
+    const std::uint32_t _sorted_elem_count = de::integral_cast<std::uint32_t>(sorted_elem_count);
+
+    init(queue_, val_in.get_dimension(0));
```

Comment:
Why do we reinitialize the object here? I'd suggest to simplify logic. Two possible solutions:
1. Initialize `buffer_` once in the constructor and check if its size matches `val_in.get_dimension(0)`, assertion is enough.
2. Do not allocated buffer in constructor, but do lazy initialization in this function.

The second is preferred to me as API prevents you from making errors.

Reply:
Probably the same issue with `radix_sort_indices_inplace`

Reply:
Agree, migrated to second solution for both.